Fix ellipseMode(CORNERS) and rectMode(CORNER)

martinleopold opened the pull request:
Resolves #7289

Changes:
This mainly resolves a rendering/crashing issues with `ellipseMode(CORNERS)` affecting `ellipse()` and `arc()`; see issue above). It also fixes a related issue with using `rectMode(CORNER)` and negative widths and/or heights.

The bulk of the changes are in `modeAdjust()` which should now properly handle the modes and return a normalized bounding box. As a result a few lines of handling negative widths and heights needed to be removed from `arc()` and `_renderEllipse()`.

I've included a test sketch in test/shape-modes.html where you can cycle through all the different modes and shapes. I'm happy to update this, if somebody could direct me on how to properly test this case. I'm not sure, since it's mostly about the rendering output.

Screenshots of the change:
![ellipse-fix](https://github.com/user-attachments/assets/d5554f2c-273f-40eb-91b3-3c528ae81e57)


#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

perminder-17 commented:
Hi, Thanks for taking up this task. I believe I have tested this and it works well. I have a concern: are you planning to write tests for it in the `test/shape-modes.html` file?

If so, I don't think that approach would work for your case. Perhaps we could write unit or visual tests instead. What's your opinion?

perminder-17 commented on test/shape-modes.html:
```diff
@@ -0,0 +1,183 @@
+            rect(x1, y1, x2, y2);
+        }
+      } catch (e) {
+        console.warn(`Error drawing (${x1}, ${y1}, ${x2}, ${y2})`);
```
If you are trying to write tests for it? You could probably refer to this file, this could help you to write tests for it? 

https://github.com/processing/p5.js/blob/main/contributor_docs/unit_testing.md

Let me know if this file is not related to testing purpose.

martinleopold commented:
Hi, I've added proper visual test and removed `test/shape-modes.html`. I had to update `test/node/helpers.js` (the test for `modeAdjust()` as well. `npm run test` passes all tests now. Please have a look and let me know if that works.

martinleopold commented:
> 🎉 Thanks for opening this pull request! Please check out our [contributing guidelines](https://github.com/processing/p5.js/blob/main/CONTRIBUTING.md) if you haven't already. And be sure to add yourself to the [list of contributors on the readme page](https://github.com/processing/p5.js#contributors)!

@all-contributors please add @martinleopold for bug, code

perminder-17 commented on test/unit/visual/cases/shape_modes.js:
```diff
@@ -0,0 +1,118 @@
+
+      for (let mode of MODES) {
+        visualTest(`Mode ${mode}`, function(p5, screenshot) {
+          p5.createCanvas(240, 500);
```
Alright, I believe creating a larger canvas might be an issue. When the tests run, a bigger canvas could make them take a bit longer due to its size. Plus, including files that are 15-20kB could make the codebase pretty heavy. Could we perhaps reduce it to 50x50? That would probably take up minimal space in bytes.

By doing this, we could include the tests one by one. For reference, you might want to check out this link: https://github.com/processing/p5.js/pull/6783/files#diff-aefa4b8e5287fbd0cdf37a6ec8d27322a4e5c893e99ec74d32053e60e2bdfd7d

Maybe by using a 50x50 size, we could adjust the parameter values of _shapeCorners accordingly?

martinleopold replied:
I think in this case using many smaller canvases instead of a few bigger ones, will actually make things slower. Because then we'll have the overhead of createCanvas() i.e. creating a rendering context much more often, and the overall codebase size will also be larger with many small PNGs because of the PNG headers.

And to thoroughly test everything, we have 4 modes, times 3 shapes, times 4 quadrants, times 4 combinations of points, equals 192 shapes to draw. I see no way around that. Bear in mind this adds 192 test for fundamental drawing operations that were previously untested.

Therefore I propose making the canvas in the tests as small as possible, while still testing all the cases. 
I have reduced the canvas size from 240x500 to 60x125 px, which reduces the total size of the generated PNGs from 223 to 12 kB.

perminder-17 replied:
sounds good!

perminder-17 reviewed (commented):
Looks great so far. Just a few bits in the testing part. Thanks a lot for your work and your patience :)

perminder-17 reviewed (approved):
Looks good to me. 

Requesting @davepagurek for a round of review since I haven't used `ellipseMode()` much, so needs another round of review. Also, very thanks for your patience @martinleopold I am really sorry for the delay on reviewing this PR, also thanks for your hardwork.

davepagurek reviewed (approved):
Thanks everyone!
